Display previews and inline images for reddit-hosted images.

mikupls opened the pull request:
Display previews and inline images for reddit-hosted images. Partially addresses #246. Also avoid division by 0 in the rendering code.

mikupls commented:
Happy new year :wave: I'm back after a long vacation :) Great to see all the improvements to libreddit in the meantime.

---

The change in this PR might seem big, but I could not detect any breakage - I was not able to figure out why we have all these SVG elements in the first place. In any case, as described in #246, it seems that reddit does not give us image dimension data, so we have to be able to live without it. I hope this PR allows us to do so.

Diff when browsing:
![Screenshot from 2022-01-09 23-11-30](https://user-images.githubusercontent.com/93015331/148703784-da0c9b8a-fe29-462f-b983-9919bf0928d3.png)

Diff when inside the post:
![Screenshot from 2022-01-09 23-12-06](https://user-images.githubusercontent.com/93015331/148703793-85ebe2de-32f8-4d2f-b294-9514f7f75539.png)

(left is the new behaviour in both cases.)

mikupls commented:
I was just catching up on new issues, and this PR also seems to fix #389 (by removing the SVG tags).

ThePixelDeveloper commented:
Regarding the SVG changes: https://github.com/spikecodes/libreddit/pull/68

It seems the author wanted to have a non jumping page when the images failed to load.

spikecodes commented:
Welcome back @mikupls! :wave:

Thank you for this PR but the removal of the SVG wrapper would cause jumping (also known as [Content Reflow](https://developers.google.com/speed/docs/insights/browser-reflow)) during image loading or perhaps image failing, as mentioned by @ThePixelDeveloper. I can almost fix this but the lack of dimensions for Reddit-hosted images in the API prevents me from entirely preventing the jumping from happening.

What is strange is that `data["preview"]` is shown when viewing Reddit while logged in but is completely hidden when viewing while signed-out. This appears to only occur for i.redd.it (such as [here](https://www.reddit.com/r/EscapefromTarkov/comments/odn91o/therapist_level_1_now_sells_lucky_scav_junk_box/.json)) images not others. :thinking:

Logged in:
![image](https://user-images.githubusercontent.com/19519553/149251018-0b9795bd-327e-4684-8dca-18d34d5934a6.png)

Logged out:
![image](https://user-images.githubusercontent.com/19519553/149251042-66f0ead8-049b-43ba-a0eb-e17ee81205be.png)

Therefore Libreddit cannot receive the dimensions and thus, can't load the Reddit-hosted images without causing content reflow. This wouldn't be an issue in individual posts as much as it would when viewing feeds that contain multiple Reddit-hosted images, causing reflowing on each load.

mikupls commented:
Thank you both for the explanations!

So what is the proposal or solution here? Only display these images in-line when inside the post? (I would prefer also having the previews, but I use my own instance, so I'm not really affected by the reflow)

If we keep the svg tags or use a similar work-around, please add a comment about the reflow considerations.

mikupls commented:
Hi :wave: 

I just rebased on master and added my old fix for backslash rewriting. I've been using both commits personally for some months now and did not encounter any problems myself.

spikecodes commented:
Hi mikupls! Just got back from my break and I'd love to accept this PR but content reflow still seems to be an issue. The SVG tags alone won't fix this as we still need to supply the image dimensions to make sure the place in held to prevent reflow. Would it be okay to only display these images in posts not feeds? The feed code could be commented out until we find a solution

mikupls commented on templates/utils.html:
```diff
@@ -104,6 +110,10 @@
 		</svg>
+		{% else %}
+		<!-- debug: simplified image thumbnail -->
+		<img loading="lazy" alt="Post image" src="{{ post.media.url }}"/>
```
Ah, I think we might have to set the width/height here explicitly in the path where we have non-zero values?

It hard for me to test content-reflow. If anyone knows how to do it reliably, let me know please.

spikecodes replied:
Yeah, if we have non-zero width and height, we should definitely use them.

To test content-reflow, I recommend refreshing with Control + F5, which clears cache so the image might take longer to load and you can see the split-second where the comments jump from the middle of the page to the bottom

spikecodes commented:
Sorry, I'm trying to understand your second work-around version. Are you thinking define a fixed space for previews without dimensions and then have the image fill vertically?

mikupls commented:
So what the current version does is:
- if we have the dimensions just use a normal img without the svg (see #389 on why it breaks things)
- if we have no dimensions, use the svg as-is. **Except**: because we do not have dimensions we have to pick some size: I just picked "100%" so it will fill whatever space there already is. we could also make the space bigger or quadratic or so. There is just no good solution that will not look bad sometimes in our case: we do not have dimensions and we do not want to wait to have them (content reflow). The only other solution is not having previews at all. (That is my understanding, maybe someone can come up with something better still)

So in summary I think the answer to you question is yes.

mikupls commented on templates/utils.html:
```diff
@@ -94,7 +94,13 @@
+		{% if post.media.height == 0 || post.media.width == 0 %}
+		<!-- debug: reddit-hosted: svg workaround image thumbnail -->
 		<svg
 			width="{{ post.media.width }}px" 
```
Need to play around with what we can put here because we do not have width/height (they are 0 in this path)

mikupls commented:
Pushed another update which should now actually match the behaviour that I described in the comment above. Sorry for the confusion.

mikupls commented:
Well it would've been too nice if things worked... My inexperience with front-end is showing.

I already found a new problem, this time with css. In the current version we seem to miss at least a `height: auto;` which is currently provided by `.post_media_image svg` (which no longer applies in my version).

I apparently also tested the content reflow incorrectly; my new approach is to zoom out to the maximum on `r/pics` and reload the page without cache. Now I see that my current version still seems to have problems (maybe due to the missing background css that was provided by the svg before...)

I'll get some sleep and maybe revisit. If you like, we can also finish this PR without changes to `utils.html` for now.

Current broken example:
![Screenshot from 2022-03-17 23-02-26](https://user-images.githubusercontent.com/93015331/158901975-7903082c-11eb-4b11-a5d4-92747cb00698.png)

mikupls commented:
Closing in favor of #738.
